05-JWT-Basics/preferred/middleware/auth.js

```diff
@@ -0,0 +1,29 @@
+const jwt = require("jsonwebtoken");
+const {UnauthenticatedError} = require("../errors");
+
+const authanticationMiddleware = async (req, res, next) => {
+    const { authorization } = req.headers;
+    console.log("Authorization", authorization);
+    console.log(req.headers);
+    if (!authorization || !authorization.startsWith("Bearer ")) {
+      throw new UnauthenticatedError("No token provided");
+    }
+
+    const token = authorization.split(" ")[1];
+    console.log(token);
+    try {
+      const decoded = jwt.verify(token, process.env.JWT_SECRET);
+      const {name} = decoded;
+      console.log(`name is: ${name}`);
+      console.log(`inside auth middleware${name}`)
+      req.user = {name}
+      next();
+    } catch (error) {
+      throw new UnauthenticatedError(
+        "Not authorizad to access this route"
+      );
+    }
+  
```

Comment:
Nitpick:
This token validation logic can be simplified with [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). That said, we don't really care if the header and token is there, instead we only really care if the header token is valid!
```js
// optional chaining allows you to access a property that may or may not be there
// if it's not there, it'll fall back to undefined
const token = req.headers.authorization?.split(" ")?.[1];

try {
  req.user = jwt.verify(token, process.env.JWT_SECRET);
  next();
} catch {
  throw new UnauthenticatedError("Not authorized");
}
```
---
Alternatively, you can skip using optional chaining and instead just try to access the property within the `try` block:
```js
try {
  // this will throw an error if the authorization header is undefined
  const token = req.headers.authorization.split(" ")[1];
  req.user = jwt.verify(token, process.env.JWT_SECRET);
  next();
} catch {
  throw new UnauthenticatedError("Not authorized");
}
```